tools/lsp/preview/element_selection.rs

```diff
@@ -0,0 +1,274 @@
+        .map(|(f, sl, sc, el, ec)| {
+            use lsp_types::{Position, Range};
+            let start = Position::new((sl as u32).saturating_sub(1), (sc as u32).saturating_sub(1));
+            let end = Position::new((el as u32).saturating_sub(1), (ec as u32).saturating_sub(1));
+
+            (f, Range::new(start, end))
+        })
+        .unwrap_or_default()
+}
+
+// triggered from the UI, running in UI thread
+pub fn element_covers_point(
+    x: f32,
+    y: f32,
+    component_instance: &ComponentInstance,
+    selected_element: &ElementRc,
+) -> bool {
+    let click_position = LogicalPoint::from_lengths(LogicalLength::new(x), LogicalLength::new(y));
+
+    let Some(position) = component_instance.element_position(selected_element) else {
+        return false;
+    };
+
+    position.contains(click_position)
+}
+
+pub fn unselect_element() {
+    super::set_selected_element(None, ComponentPositions::default());
+    return;
+}
+
+fn select_element(
+    x: f32,
+    y: f32,
+    component_instance: &ComponentInstance,
+    selected_element: Option<&ElementRc>,
+) -> Option<ElementRc> {
+    let click_position = LogicalPoint::from_lengths(LogicalLength::new(x), LogicalLength::new(y));
+
+    let Some(c) = selected_element else {
+        unselect_element();
+        return None;
+    };
+
+    let Some(position) = component_instance.element_position(&c) else {
+        return None;
+    };
+    if position.contains(click_position) {
+        let secondary_positions = if let Some((path, offset)) = element_offset(&c) {
+            component_instance.component_positions(path, offset)
+        } else {
+            ComponentPositions::default()
+        };
+
+        super::set_selected_element(Some((&c, position)), secondary_positions);
+        let document_position = lsp_element_position(&c);
+        if !document_position.0.is_empty() {
+            super::ask_editor_to_show_document(document_position.0, document_position.1);
+        }
+        return Some(c.clone());
+    }
+
+    None
+}
+
+// triggered from the UI, running in UI thread
+pub fn select_element_at_impl(
+    x: f32,
+    y: f32,
+    component_instance: &ComponentInstance,
+    root_element: &ElementRc,
+    current_element: Option<&ElementRc>,
+    reverse: bool,
+) -> Option<ElementRc> {
+    let re = root_element.borrow();
+    let mut fw_iter = re.children.iter();
+    let mut bw_iter = re.children.iter().rev();
+
+    let iterator: &mut dyn Iterator<Item = &ElementRc> =
+        if reverse { &mut bw_iter } else { &mut fw_iter };
+
+    let mut skip = current_element.is_some();
+    for c in &mut *iterator {
+        let c = self_or_embedded_component_root(c);
+
+        if skip {
+            if let Some(ce) = current_element {
+                if Rc::ptr_eq(ce, &c) {
+                    skip = false;
+                }
+            }
+            continue;
+        }
+
+        if let Some(result) = select_element(x, y, component_instance, Some(&c)) {
```

Comment:
imho you should try to recurse here and try to find the top element on the first click